Updates of SDIRK and generic implicit integrators

devmotion opened the pull request:
Again this updates calculation of error estimates and function arguments. However, this PR might need a more careful inspection since I also tried to reduce the number of calculations and therefore I introduced some new variables, especially in TRBDF2. The SDIRK integrators could be further improved, I guess, by saving `κ*tol` instead of both `κ` and `tol` in the cache (at least in `perform_steps!` only the product is needed) and by saving all constants of TRPBDF2 in a tableau. This is not part of this PR yet, but could be made an additional commit. I hope I didn't introduce any floating point issues by these simplifications.

This PR should hopefully also address https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/134, however, I'm not sure whether https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/139 still remains an issue.

I compared these updated methods to the master branch with the ROBER benchmark for high tolerances:

master
![figure_1_sdirk_master](https://user-images.githubusercontent.com/26102/29173559-f7542ec2-7de3-11e7-95bb-a80308a6f1e3.png)

PR
![figure_1_sdirk_pr](https://user-images.githubusercontent.com/26102/29173570-01df2ed2-7de4-11e7-93cd-4bdd4aff476a.png)

The plots are very similar apart from the Trapezoid curves which I don't understand.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -1,27 +1,29 @@
+  @unpack t,dt,uprev,u,f = integrator
+  @unpack uf,κ,tol = cache
+
+  κtol = κ*tol # save in cache instead of κ and tol?
```
it's fine here. No need to over optimize. This allows for making the values adaptive in the future.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -72,15 +67,21 @@ end
   cache.ηold = η
   cache.newton_iters = iter
-  u = uprev + z
+  u = u .+ dz
```
should be `u = uprev .+ dz`

devmotion replied:
Because of floating point issues otherwise?

devmotion replied:
Since mathematically we have `z <- u - uprev` initially, then an update `z <- z + dz`, at the beginning of every iteration `u <- u + dz = u + z`, at the end of every iteration `z <- z + dz`, and hence at this point (since either the loop was completely skipped or left after a full iteration - premature exit of an iteration by failed convergence is handled before) again `u <- u + dz = uprev + z`.

ChrisRackauckas replied:
If floating point works correctly then yes. There's a good chance this adds a few small values in the end of the Newton iteration though if the equation is stiff enough and the values are large enough. In that case, repeat adding of `dz` to `u` will accumulate errors. It's fine to "accumulate errors` onto `z` because that's the iteration variable, so it will iterate more if `z` is sufficiently incorrect. But accumulating into `u` which is then called in `f` can lead to `f` being off, which will cause problems in stiff problems. In the end, this accumulation doesn't make `u` exactly satisfy the tableau which can lead to order loss.

devmotion replied:
Yes, that makes sense. But we could e.g. still cache the result of `uprev + ω*zprev + ω*zᵧ` and replace the brackets in lines https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/sdirk_integrators.jl#L767 and
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/sdirk_integrators.jl#L792
with it, couldn't we? We could add `d*z` to this result in every iteration and after the loop to obtain the next `u`. But we should not just increase `u` by some `d*Δz` every time because this would accumulate errors. Is that correct?

ChrisRackauckas replied:
Yes, that's actually precisely what Hairer does.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -304,10 +287,7 @@ end
-  integrator.fsallast = f(t+dt,u)
-  integrator.k[1] = integrator.fsalfirst
-  integrator.k[2] = integrator.fsallast
+  u = @. u + (z + dz)
```
This isn't right. `u = uprev + 2z`

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -405,18 +395,14 @@ end
   fail_convergence = false
   while (do_newton || iter < integrator.alg.min_newton_iter) && iter < integrator.alg.max_newton_iter
     iter += 1
+    u .+= dz
```
This isn't right. `@. u = uprev + z`

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -230,71 +218,66 @@ end
   while (do_newton || iter < integrator.alg.min_newton_iter) && iter < integrator.alg.max_newton_iter
     iter += 1
-    b = -z .+ dto2.*f(t+dto2,uprev + z)
+    u = u .+ dz
```
This isn't right. `u = uprev .+ z`

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -209,19 +196,20 @@ end
   cache.ηold = η
   cache.newton_iters = iter
-  @. u = uprev + z
+  u .+= dz
```
This isn't right.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -169,18 +161,14 @@ end
   fail_convergence = false
   while (do_newton || iter < integrator.alg.min_newton_iter) && iter < integrator.alg.max_newton_iter
     iter += 1
+    u .+= dz
```
This isn't right.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -481,204 +465,211 @@ end
   zᵧ = zprev
   iter = 1
-  uᵧ = @. (uprev + d*zprev) + d*zᵧ
+  uᵧ = @. uprev + γ*zᵧ
```
Add a comment for this.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -481,204 +465,211 @@ end
 
   iter = 1
-  u = @. (uprev + ω*zprev + ω*zᵧ) + d*z
+  u = @. uprev + c1*zprev + c2*zᵧ
```
Add a comment here.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -687,11 +678,14 @@ end
   iter = 1
-
-  @. uᵧ = (uprev + d*zprev) + d*zᵧ
+  @. uᵧ = uprev + γ*zprev
```
Add a comment here.

ChrisRackauckas replied:
Because this would be wrong if changed to extrapolation.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -687,11 +678,14 @@ end
     end
   end
 
+  # TODO: Add extrapolant initial guess
```
No, this guess is the smoothed aggregation form FSAL, this should not be extrapolated.

devmotion replied:
Ah OK, I just copied it from the not in-place method. So I'll remove it in both the in-place and not in-place method?

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -687,11 +678,14 @@ end
+
+  ##### Solve Trapezoid Step
+
   @. zᵧ = zprev
```
This could possibly be changed to an extrapolation in the future.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -481,204 +465,211 @@ end
 
   ##### Solve Trapezoid Step
 
   zᵧ = zprev
```
As with below, the extrapolant comment should be moved down here. And it shouldn't be assumed this will hold.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -445,7 +430,7 @@ end
   cache.ηold = η
   cache.newton_iters = iter
-  @. u = uprev + 2*z
+  @. u += z + dz
```
I don't think this is right? `u` is aggregated differently here than in the stages. It's tough to understand if this is correct.

devmotion replied:
Mathematically it should be correct, I guess, but again it is apparently not a good idea because it could accumulate floating point errors.

ChrisRackauckas commented:
In the SDIRK methods there is an aggregation by which the `u`s are defined via the `z`s in a transform. It's done like this so that way they don't incur floating point issues due to successive iterations. Then after the iterations, the recombination of the last `u` is not necessarily the same as the one used in the iterations. That is definitely the case the `Trapezoid`, and probably what caused the issue there. While the other way would give you second order, it would not always work well on stiff equations, which is what you see in the ROBER benchmark (thanks for showing that sanity check).

devmotion commented:
I corrected the update of `u` in these methods, but the strange result for the trapezoidal rule was not affected by this change. However, I figured out that this is due to error estimation with divided differences which I tried to simplify; but apparently I changed it in such a way that it is more vulnerable to floating point issues. It seems divided differences are always vulnerable to roundoff errors when computed in a recursive fashion, so maybe it would be good to improve it e.g. following the notes on page 191 of http://profs.scienze.univr.it/~caliari/pdf/C07.pdf or any other better algorithm?

And I was wondering why we actually divide by 12 in this case - I know the local truncation error can be bound by `h³/12*max |y'''(t)|`, but according to the [mean value theorem for divided differences](https://en.wikipedia.org/wiki/Mean_value_theorem_(divided_differences)) the 3rd divided difference that we calculate already equals `y'''(s)/6` for some `s` in the time interval from `tprev2` to `t+dt`. So why do we not just divide the resulting 3rd divided difference by 2?

ChrisRackauckas commented:
>And I was wondering why we actually divide by 12 in this case - I know the local truncation error can be bound by h³/12*max |y'''(t)|, but according to the mean value theorem for divided differences the 3rd divided difference that we calculate already equals y'''(s)/6 for some s in the time interval from tprev2 to t+dt. So why do we not just divide the resulting 3rd divided difference by 2?

Yeah, that looks like it's a mistake. See "The Spice Book" for the original reference. I looked again and they used 7/12, but note that the DD can overstate the truncation error by several times.

>It seems divided differences are always vulnerable to roundoff errors when computed in a recursive fashion, so maybe it would be good to improve it e.g. following the notes on page 191 of http://profs.scienze.univr.it/~caliari/pdf/C07.pdf or any other better algorithm?

Yeah, it's a problem. This should be a separate PR though.

devmotion commented:
I corrected the Newton iterations and added some comments. Moreover, I added a correction factor and set it to the default value `7/12` of the SPICE book. However, in my opinion the calculation of divided differences seems not only vulnerable to rounding errors but also mathematically incorrect in the master branch - according to the definition of divided differences there should only be recursive subtractions and divisions, I guess.

So I changed the calculation of 2nd and 3rd divided differences in both Euler and Trapezoid methods to a hopefully correct recursive calculation, i.e. I mainly replaced some `+` signs with `-` signs compared to the master branch. The convergence tests still pass but this leads to a really bad performance of the Trapezoid methods in the ROBER benchmarks whereas the Euler methods are almost unaffected by this change. I assume this is due to the fact that the new introduced subtractions have a much larger impact on the calculation of higher order divided differences. I tried a lot of things to improve the benchmarks again, but neither changing broadcasts to for loops nor defining `DD30` and `DD31` in the same style with only replaced `-` signs had any influence. However, since I'm far from an expert in both ODE solvers and divided differences maybe I messed up something and the original version was correct?

PR
![figure_1_pr_last_default_spice](https://user-images.githubusercontent.com/26102/29226128-2b62512e-7ed1-11e7-8d56-539570dfb28e.png)

I also did some tests with changing the correction factor to 1/6 (current master) or 1 (i.e. without any additional scaling):

1/6:
![figure_1_pr_1-6](https://user-images.githubusercontent.com/26102/29226383-26b7a9f2-7ed2-11e7-9aa5-0c302fa85c61.png)

1:
![figure_1_pr_1](https://user-images.githubusercontent.com/26102/29226386-2cc5df44-7ed2-11e7-87ff-04d3df807193.png)

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -390,11 +407,8 @@ end
-
-  iter += 1
+  iter = 1
+  @. u = uprev + z
```
Redundent

devmotion replied:
I thought so in the beginning as well, but then I got confused because of lines such as https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/141/files/660aa9c90743d54f2aa6c22468ea561ba3480cf9?diff=split#diff-57e331c13262e0ef1f111ed17cda714fL39 where exactly this recalculation of `u` (which might be different to the old `u` because of rounding errors, I guess?) is done before the function evaluation - but also it was only present in the not in-place method... So in the end I decided to do these recalculations in every method. But if it's not needed, I'll just remove it everywhere.

devmotion replied:
By the way, the sam line is also present in the new SDIRK2 method: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/sdirk_integrators.jl#L855

ChrisRackauckas replied:
Yeah it was a mistake. These implementations haven't been cleaned up yet so there's some details that I've opened issues for to investigate. This is one of those things I noticed that can probably be removed, but didn't because I'd want to regression test when doing so.

ChrisRackauckas commented on src/integrators/sdirk_integrators.jl:
```diff
@@ -725,30 +754,36 @@ end
+  with d = (2 - sqrt(2))/2 can be simplified to
+    @. z = -sqrt(2)/2*zprev + (1 + sqrt(2)/2)*zᵧ
+  =#
+  @. z = a1*zprev + a2*zᵧ
```
You double checked that this did not change the predictions?

devmotion replied:
You mean whether the new definition leads to the same `z` as the old one? Or whether the solution stays the same? Is there a reason to prefer the old version to the mathematically equivalent new version if there would be a slight difference (due to floating point issues maybe)?

ChrisRackauckas commented:
I really really appreciate you doing this. However, since these algorithms are very new and in the cleaning phase I can't accept a PR without showing no regressions since I'm still only barely sure it all works correctly (for example, I'm pretty sure the extrapolation timing in Trapezoid should be dt/2, which hasn't been done yet). So I think you should split the PR to do all of the aesthetic and broadcasting fixes so that can merge quickly, and keep the divided differences change separate.

The other thing to note is that Trapezoid is in a different scaling for `z` than the other algorithms. It's using the Hairer transform. I realized later I didn't like it, but haven't gotten back around to it. But this may be throwing off the reasoning about the algorithm since it is different than the others.

devmotion commented:
I can completely understand that, and I'll just transfer all other changes except of the divided differences to a new PR. Nevertheless, I think there might be both a mathematical and a numerical problem with these divided differences, so one should keep it in mind at least and have another look at it when everything else is settled down.

ChrisRackauckas commented:
>Nevertheless, I think there might be both a mathematical and a numerical problem with these divided differences, so one should keep it in mind at least and have another look at it when everything else is settled down.

Agreed.

ChrisRackauckas commented:
`sdirk_integrators.jl` changed quite substantially. Do you plan on reviving this? Or splitting the efforts to new PRs?

devmotion commented:
I'm already working on a new PR without the changes of divided differences, just have to do some final tests and corrections. So I'll close this and will open another PR soon which will already cover the changes in `sdirk_integrators.jl`.
